FLT/AutomorphicRepresentation/Example.lean

```diff
@@ -482,8 +482,139 @@ lemma unitsrat_meet_unitszHat : unitsratsub ⊓ unitszHatsub = unitszsub := by
+lemma unitsrat_join_unitszHat : unitsratsub ⊔ unitszHatsub = ⊤ := by
+  apply le_antisymm
+  · apply le_top
+  · intro y _
+    rcases canonicalForm y.val with ⟨N, x, hy⟩
+    rcases canonicalForm (y⁻¹.val) with ⟨N2, x2, hy2⟩
+    set xinv := (1 / (N * N2) : ℚ) ⊗ₜ[ℤ] x2 with xinv_def
+    have : (i₂ x) * xinv = 1 := by
+      rw [xinv_def, Algebra.TensorProduct.includeRight_apply, one_div, mul_inv_rev,
+        Algebra.TensorProduct.tmul_mul_tmul,one_mul,mul_comm,← Algebra.TensorProduct.tmul_mul_tmul,
+        ← one_div, ← one_div, ← hy, ← hy2, ← Units.val_mul, mul_inv_cancel, Units.val_one]
+    let xunit : QHatˣ := ⟨i₂ x, xinv, this, by rw [mul_comm]; exact this⟩
+    suffices h : ∀ (u : QHatˣ), (u : QHat) ∈ zHatsub → u ∈ unitsratsub ⊔ unitszHatsub by
+      specialize h xunit
+      simp only [Algebra.TensorProduct.includeRight_apply, AddMonoidHom.mem_range,
+        AddMonoidHom.coe_coe, exists_apply_eq_apply, forall_const, Subgroup.mem_sup, xunit] at h
+      rcases h with ⟨w, ⟨v, hv⟩, z, ⟨t, ht⟩, wzx⟩
+      rw [Subgroup.mem_sup]
+      let q : ℚˣ := ⟨v / N, N / v, by field_simp, by field_simp⟩
+      use ((Units.map ↑i₁) q)
+      simp only [MonoidHom.mem_range, exists_exists_eq_and]
+      refine ⟨⟨q, rfl⟩, t, ?_⟩
+      simp only [← Units.eq_iff, hy, ht, Units.map_mk, MonoidHom.coe_coe,
+        Algebra.TensorProduct.includeLeft_apply, Units.val_mul, one_div, q, xunit]
+      rw [← mul_one (N⁻¹ : ℚ), ← one_mul x, ← Algebra.TensorProduct.tmul_mul_tmul, div_eq_mul_inv,
+        mul_comm (v : ℚ), ← mul_one 1, ← Algebra.TensorProduct.tmul_mul_tmul, mul_assoc, mul_one]
+      congr
+      simpa only [← hv, ← Units.eq_iff, Units.val_mul, Units.coe_map, MonoidHom.coe_coe,
+        Algebra.TensorProduct.includeLeft_apply, xunit, q] using wzx
+    clear * -
+    intro x hx
+    rcases canonicalForm (x⁻¹.val) with ⟨M, y, hxinv⟩
+    have : x * (i₂ y) = M := by
+      rw [← one_mul (M : QHat), ← Units.val_one, ← mul_inv_cancel x, Units.val_mul, mul_assoc]
+      congr!
+      have h : (M : QHat) = (M : ℚ) ⊗ₜ[ℤ] 1 := by norm_cast
+      rw [Algebra.TensorProduct.includeRight_apply, hxinv, h, Algebra.TensorProduct.tmul_mul_tmul,
+        mul_one, one_div, inv_mul_cancel₀]
+      simp only [ne_eq, Rat.natCast_eq_zero, PNat.ne_zero, not_false_eq_true]
+    rcases hx with ⟨X, hX⟩
+    let I := Ideal.span {X}
+    let J := (Int.castRingHom ZHat) ⁻¹' I -- this needs to be an ideal
+    have Jnonzero : (M : ℤ) ∈ J := by
+      simp only [Int.coe_castRingHom, Set.mem_preimage, Int.cast_natCast, SetLike.mem_coe, J, I]
+      rw [Ideal.mem_span_singleton']
+      use y
+      apply injective_zHat
+      simp only [mul_comm, ← hX, AddMonoidHom.coe_coe, Algebra.TensorProduct.includeRight_apply,
+        Algebra.TensorProduct.tmul_mul_tmul, mul_one, I, J] at this
+      rw [Algebra.TensorProduct.includeRight_apply, this, map_natCast]
+    let IdealJ : Ideal ℤ := by -- make this construction into a function (for Mathlib)?
```

Comment:
All basic stuff like this should already be there. Your question shouldn't be "should I put preimages of ideals in mathlib" but rather "how do I find preimages of ideals in mathlib". It's helpful to know buzzwords, like `comap` (which we use for preimage, pullback etc).